lib/Conversion/StandardToHandshake/StandardToHandshake.cpp

```diff
@@ -614,6 +615,191 @@ static int getBranchCount(Value val, Block *block) {
+  // assumes that merge block insertion happended beforehand
+  // Thus, for each split block, there exists one merge block which is the post
+  // dominator of the child nodes.
+  Region &r = hl.getRegion();
+  Operation *parentOp = r.getParentOp();
+
+  // Assumes that each loop has only one exit block. Such an error should
+  // already be reported by the loop rewriting.
+  assert(loopsHaveSingleExit(loopAnalysis) &&
+         "expected loop to only have one exit block.");
+
+  // TODO: cannot handle irreducible control flow at this point.
+
+  for (Block &b : r) {
+    if (b.getNumSuccessors() < 2)
+      continue;
+
+    // Loop headers cannot be merge blocks.
+    if (loopAnalysis.isLoopElement(&b))
+      continue;
+
+    assert(b.getNumSuccessors() == 2);
+    Block *succ0 = b.getSuccessor(0);
+    Block *succ1 = b.getSuccessor(1);
+
+    if (succ0 == succ1)
+      continue;
+
+    Block *mergeBlock = postDomInfo.findNearestCommonDominator(succ0, succ1);
+
+    // Precondition checks
+    if (formsIrreducibleCF(&b, mergeBlock))
+      return parentOp->emitError("expected only reducible control flow.");
+
+    unsigned nonLoopPreds = 0;
+    LoopInfo *info = loopAnalysis.getLoopInfoForHeader(mergeBlock);
+    for (auto pred : mergeBlock->getPredecessors()) {
+      if (info && info->inLoop.contains(pred))
+        continue;
+      nonLoopPreds++;
+    }
+    if (nonLoopPreds > 2)
+      return parentOp->emitError(
+          "expected a merge block to have two predecessors. Did you run the "
+          "merge block insertion pass?");
+
+    blockPairs.emplace_back(&b, mergeBlock);
+  }
+
+  return success();
+}
+
+LogicalResult FeedForwardNetworkRewriter::apply() {
+  if (failed(loopAnalysis.analyzeRegion()))
+    return failure();
+
+  BlockPairs pairs;
+  if (failed(findBlockPairs(pairs)))
+    return failure();
+
+  for (auto [splitBlock, mergeBlock] : pairs) {
+    auto buffer = buildSplitNetwork(splitBlock);
+    buildMergeNetwork(mergeBlock, buffer);
+  }
+
+  return success();
+}
+
+BranchOp FeedForwardNetworkRewriter::buildSplitNetwork(Block *splitBlock) {
+  SmallVector<handshake::ConditionalBranchOp> branches;
+  llvm::copy(splitBlock->getOps<handshake::ConditionalBranchOp>(),
+             std::back_inserter(branches));
+
+  Value cond = branches[0].conditionOperand();
+  assert(llvm::all_of(branches, [&](auto branch) {
+    return branch.conditionOperand() == cond;
+  }));
+
+  rewriter.setInsertionPointAfterValue(cond);
+  // Requre a cast to index to stick to the type of the mux input.
+  Value condAsIndex = rewriter.create<arith::IndexCastOp>(
+      cond.getLoc(), rewriter.getIndexType(), cond);
+
+  // TODO how to size these?
+  // Longest path in a CFG-DAG would be O(#blocks)
+  auto buffer = rewriter.create<handshake::BufferOp>(
+      cond.getLoc(), rewriter.getIndexType(), /*size=*/2, condAsIndex,
+      /*bufferType=*/BufferTypeEnum::fifo);
+
+  return rewriter.create<handshake::BranchOp>(cond.getLoc(), buffer);
```

Comment:
Does this branch even do anything meaningful? The buffer not being part of a CFG "Block", a branching operation does not seem to have any meaning.

Reply:
Handshake branch ops should in general have some thought poured into them... As far as I know, they're left over from the Dynamatic reimplementation. Branch ops don't do anything, and are always removed through canonicalization! https://github.com/llvm/circt/blob/6515ddcb70e11fca5f1e277a95ee263cf2391712/include/circt/Dialect/Handshake/HandshakeCanonicalization.td#L25-L26

Reply:
I saw that they are lowered to identity circuits, yeah. 
Then we do not require one here, I guess.